pyomo/dae/flatten.py

```diff
@@ -7,147 +7,371 @@
+        if val is _NotAnIndex:
+            filled_index[i] = index[j]
+            # `index` is always a tuple, so this is valid
+            j += 1
+    # Make sure `filled_index` had the same number of vacancies
+    # as `product` has factors. Not _strictly_ necessary.
+    assert j == len(index)
+    filled_index = tuple(filled_index)
+
+    if len(filled_index) == 1:
+        return filled_index[0]
+    else:
+        return filled_index
+
+def _fill_indices_from_product(partial_index_list, product):
+    """ 
+    `partial_index_list` is a list of indices, each corresponding to a
+    set. If an entry in `partial_index_list` is `_NotAnIndex`, that
+    slot will get filled in by an entry from `setprod`.
+
+    `product` is a `SetProduct` with as many "factors" as there are
+    missing indices in `partial_index_list`.
+    """
+    # We will manipulate `normalize_index.flatten`.
+    # Store its original value so we can reset it when we're done.
+    _normalize_index_flatten = normalize_index.flatten
+    try:
+        normalize_index.flatten = False
+        for index in product:
+            # Since `normalize_index.flatten` is False, `index` is a
+            # scalar or (tuple of (scalars or tuples)). Conveniently,
+            # each entry in the tuple belongs to a single factor set.
+            # I do not have to worry about having a product-of-products
+            # here because I created the product from "unfactorable sets"
+
+            # We need to generate a new index for every entry of `product`,
+            # and want to reuse `partial_index_list` as a starting point,
+            # so we copy it here.
+            filled_index = list(partial_index_list)
+
+            normalize_index.flatten = _normalize_index_flatten
+
+            # filled_index can now be used in the user's intended way
+            # This determines how we will access the component's data
+            # with our new index, which is currently _completely_ unflattened
+            # (i.e. a tuple-of-tuples, no further nesting).
+            #
+            # This will not work well if the user's component is indexed by
+            # a nested product of sets AND normalize_index.flatten is False.
+            # In this case we may try to access a component's data with
+            # >>> comp[(1,'a',1)]
+            # (each coordinate belongs to its own set) when it expects
+            # >>> comp[((1,'a'),1)]
+            # because `comp` was created with two set arguments, the first
+            # of which was already a product.
+
+            yield _fill_indices(filled_index, index)
+
+            normalize_index.flatten = False
+            # Want to get the unflattened factors when we advance the
+            # iterator of `product`
+    finally:
+        # Reset `normalize_index.flatten`
+        normalize_index.flatten = _normalize_index_flatten
+
+def generate_sliced_components(b, index_stack, _slice, sets, ctype, index_map):
+    """
```

Comment:
This docstring is missing a description of the `sets` argument

Reply:
Added.